File: QuantaraCore/Sources/QuantaraCore/Engine/Categorizer.swift
import Foundation

public struct CategorizationResult: Hashable, Sendable {
    public let category: ExpenseCategory
    /// Confiance dans [0, 1]. En dessous de 0,6 l'app propose sans appliquer.
    public let confidence: Decimal
    public let matchedPattern: String?
    public let isSubscription: Bool
}

/// Règle de catégorisation apprise ou fournie.
public struct CategorizationRule: Hashable, Codable, Sendable, Identifiable {
    public let id: UUID
    /// Motif recherché dans le libellé normalisé.
    public let pattern: String
    public let category: ExpenseCategory
    public let isSubscription: Bool
    /// Une règle créée par l'utilisateur prime sur les règles embarquées.
    public let isUserDefined: Bool

    public init(
        id: UUID = UUID(),
        pattern: String,
        category: ExpenseCategory,
        isSubscription: Bool = false,
        isUserDefined: Bool = false
    ) {
        self.id = id
        self.pattern = pattern
        self.category = category
        self.isSubscription = isSubscription
        self.isUserDefined = isUserDefined
    }
}

/// Catégorisation locale des libellés bancaires (§20).
///
/// Premier étage, entièrement hors ligne : rapide, gratuit, et surtout aucune donnée
/// n'est transmise. Le repli IA (second étage) n'intervient que sur les libellés non
/// reconnus, et uniquement si l'utilisateur a activé le mode détaillé.
public enum Categorizer {

    /// Normalise un libellé bancaire : majuscules, sans accents, sans dates, sans
    /// numéros de carte, sans mentions techniques (« CB », « PAIEMENT », « VIR »…).
    public static func normalize(_ label: String) -> String {
        let folded = label
            .folding(options: [.diacriticInsensitive, .caseInsensitive], locale: Locale(identifier: "fr_FR"))
            .uppercased()

        let noise = [
            "CARTE", "CB", "PAIEMENT", "PAIMT", "ACHAT", "VIREMENT", "VIR", "PRLV",
            "PRELEVEMENT", "SEPA", "FACTURE", "RETRAIT", "DAB"
        ]

        var cleaned = folded
        for token in noise {
            cleaned = cleaned.replacingOccurrences(of: token, with: " ")
        }

        // Supprime les nombres (dates, numéros de transaction) et la ponctuation.
        cleaned = cleaned.map { character -> Character in
            if character.isLetter || character == " " { return character }
            return " "
        }
        .reduce(into: "") { $0.append($1) }

        return cleaned
            .split(separator: " ")
            .filter { $0.count > 1 }
            .joined(separator: " ")
            .trimmingCharacters(in: .whitespaces)
    }

    /// Catégorise un libellé. Les règles utilisateur passent avant les règles embarquées :
    /// une correction manuelle doit toujours gagner.
    public static func categorize(
        label: String,
        userRules: [CategorizationRule] = []
    ) -> CategorizationResult? {
        let normalized = normalize(label)
        guard !normalized.isEmpty else { return nil }

        for rule in userRules where normalized.contains(rule.pattern) {
            return CategorizationResult(
                category: rule.category,
                confidence: 1,
                matchedPattern: rule.pattern,
                isSubscription: rule.isSubscription
            )
        }

        for rule in builtInRules where normalized.contains(rule.pattern) {
            return CategorizationResult(
                category: rule.category,
                confidence: Decimal(string: "0.85") ?? 0,
                matchedPattern: rule.pattern,
                isSubscription: rule.isSubscription
            )
        }

        return nil
    }

    /// Détecte les récurrences dans un historique de transactions.
    ///
    /// Une dépense qui revient au moins trois fois, à intervalle régulier et pour un
    /// montant stable, est presque toujours une charge fixe non déclarée.
    public static func detectRecurrences(
        in transactions: [Transaction],
        calendar: Calendar = .gregorianUTC
    ) -> [(label: String, averageAmount: Money, occurrences: Int, suggestedFrequency: Frequency)] {

        let expenses = transactions.filter { $0.kind == .expense }
        let grouped = Dictionary(grouping: expenses) { normalize($0.label) }

        return grouped.compactMap { label, group -> (String, Money, Int, Frequency)? in
            guard group.count >= 3, let currency = group.first?.amount.currency else { return nil }

            let amounts = group.map(\.effectiveAmount.amount)
            guard let mean = DecimalStatistics.mean(amounts), mean > .zero else { return nil }

            // Montant stable : écart-type inférieur à 15 % de la moyenne.
            if let deviation = DecimalStatistics.standardDeviation(amounts),
               deviation / mean > (Decimal(string: "0.15") ?? 0) {
                return nil
            }

            let dates = group.map(\.date).sorted()
            let intervals = zip(dates, dates.dropFirst()).map { earlier, later in
                calendar.dateComponents([.day], from: earlier, to: later).day ?? 0
            }
            guard !intervals.isEmpty else { return nil }
            let averageInterval = intervals.reduce(0, +) / intervals.count

            let frequency = closestFrequency(toDayInterval: averageInterval)
            guard frequency != .oneOff else { return nil }

            return (label, Money(mean, currency), group.count, frequency)
        }
        .sorted { $0.1 > $1.1 }
    }

    private static func closestFrequency(toDayInterval interval: Int) -> Frequency {
        let candidates: [Frequency] = [.weekly, .biweekly, .monthly, .quarterly, .semiannual, .annual]
        var best: Frequency = .oneOff
        var bestDistance = Int.max
        for candidate in candidates {
            guard let reference = candidate.approximateDayInterval else { continue }
            let distance = abs(reference - interval)
            // Tolérance de 20 % : un prélèvement « mensuel » tombe entre 28 et 33 jours.
            guard distance <= max(3, reference / 5) else { continue }
            if distance < bestDistance {
                bestDistance = distance
                best = candidate
            }
        }
        return best
    }

    // MARK: - Règles embarquées

    /// Dictionnaire de marchands courants en France et en Belgique.
    /// Volontairement conservateur : mieux vaut ne pas catégoriser que mal catégoriser,
    /// une erreur silencieuse fausse le budget sans que l'utilisateur s'en aperçoive.
    public static let builtInRules: [CategorizationRule] = [
        // Courses
        rule("CARREFOUR", .variable(.groceries)),
        rule("LECLERC", .variable(.groceries)),
        rule("INTERMARCHE", .variable(.groceries)),
        rule("AUCHAN", .variable(.groceries)),
        rule("LIDL", .variable(.groceries)),
        rule("ALDI", .variable(.groceries)),
        rule("MONOPRIX", .variable(.groceries)),
        rule("FRANPRIX", .variable(.groceries)),
        rule("CASINO", .variable(.groceries)),
        rule("SUPER U", .variable(.groceries)),
        rule("COLRUYT", .variable(.groceries)),
        rule("DELHAIZE", .variable(.groceries)),
        rule("PICARD", .variable(.groceries)),
        rule("BIOCOOP", .variable(.groceries)),

        // Énergie facturée par un pétrolier : placé avant « TOTAL » car les règles sont
        // évaluées dans l'ordre et la première qui correspond gagne.
        rule("TOTALENERGIES ELEC", .fixed(.electricity)),
        rule("TOTALENERGIES GAZ", .fixed(.gas)),

        // Carburant
        rule("TOTAL", .variable(.fuel)),
        rule("SHELL", .variable(.fuel)),
        rule("ESSO", .variable(.fuel)),
        rule("BP ", .variable(.fuel)),
        rule("STATION", .variable(.fuel)),

        // Transports
        rule("SNCF", .variable(.transport)),
        rule("RATP", .variable(.transport)),
        rule("NAVIGO", .variable(.transport)),
        rule("UBER", .variable(.transport)),
        rule("BLABLACAR", .variable(.transport)),
        rule("STIB", .variable(.transport)),
        rule("SNCB", .variable(.transport)),

        // Restaurants
        rule("MCDONALD", .variable(.restaurants)),
        rule("BURGER KING", .variable(.restaurants)),
        rule("DELIVEROO", .variable(.restaurants)),
        rule("UBER EATS", .variable(.restaurants)),
        rule("JUST EAT", .variable(.restaurants)),
        rule("RESTAURANT", .variable(.restaurants)),
        rule("BOULANGERIE", .variable(.restaurants)),
        rule("STARBUCKS", .variable(.restaurants)),

        // Abonnements
        rule("NETFLIX", .fixed(.subscriptions), isSubscription: true),
        rule("SPOTIFY", .fixed(.subscriptions), isSubscription: true),
        rule("DISNEY", .fixed(.subscriptions), isSubscription: true),
        rule("APPLE COM BILL", .fixed(.subscriptions), isSubscription: true),
        rule("AMAZON PRIME", .fixed(.subscriptions), isSubscription: true),
        rule("CANAL", .fixed(.subscriptions), isSubscription: true),
        rule("DEEZER", .fixed(.subscriptions), isSubscription: true),
        rule("YOUTUBE PREMIUM", .fixed(.subscriptions), isSubscription: true),

        // Télécoms
        rule("ORANGE", .fixed(.phone)),
        rule("SFR", .fixed(.phone)),
        rule("BOUYGUES", .fixed(.phone)),
        rule("FREE MOBILE", .fixed(.phone)),
        rule("PROXIMUS", .fixed(.phone)),
        rule("FREE", .fixed(.internet)),

        // Énergie et eau
        rule("EDF", .fixed(.electricity)),
        rule("ENGIE", .fixed(.gas)),
        rule("VEOLIA", .fixed(.water)),
        rule("SUEZ", .fixed(.water)),

        // Assurances
        rule("AXA", .fixed(.homeInsurance)),
        rule("MAIF", .fixed(.homeInsurance)),
        rule("MACIF", .fixed(.homeInsurance)),
        rule("MATMUT", .fixed(.homeInsurance)),
        rule("ALLIANZ", .fixed(.homeInsurance)),
        rule("MUTUELLE", .fixed(.healthInsurance)),
        rule("HARMONIE", .fixed(.healthInsurance)),

        // Santé
        rule("PHARMACIE", .variable(.health)),
        rule("LABORATOIRE", .variable(.health)),
        rule("OPTIQUE", .variable(.health)),

        // Shopping et loisirs
        rule("AMAZON", .variable(.shopping)),
        rule("FNAC", .variable(.shopping)),
        rule("DECATHLON", .variable(.shopping)),
        rule("ZARA", .variable(.shopping)),
        rule("IKEA", .variable(.shopping)),
        rule("LEROY MERLIN", .variable(.shopping)),
        rule("STEAM", .variable(.videoGames)),
        rule("PLAYSTATION", .variable(.videoGames)),
        rule("NINTENDO", .variable(.videoGames)),
        rule("XBOX", .variable(.videoGames)),
        rule("CINEMA", .variable(.leisure)),
        rule("UGC", .variable(.leisure)),
        rule("GAUMONT", .variable(.leisure)),

        // Voyages
        rule("BOOKING", .variable(.travel)),
        rule("AIRBNB", .variable(.travel)),
        rule("AIR FRANCE", .variable(.travel)),
        rule("RYANAIR", .variable(.travel)),
        rule("EASYJET", .variable(.travel)),

        // Animaux
        rule("VETERINAIRE", .variable(.pets)),
        rule("ANIMALERIE", .variable(.pets))
    ]

    private static func rule(
        _ pattern: String,
        _ category: ExpenseCategory,
        isSubscription: Bool = false
    ) -> CategorizationRule {
        CategorizationRule(
            pattern: pattern,
            category: category,
            isSubscription: isSubscription,
            isUserDefined: false
        )
    }
}

File: desktop/src/core/engine/categorizer.ts
import { Money } from '../money';
import type { ExpenseCategoryId } from '../categories';
import type { Frequency } from '../frequency';
import type { Transaction } from '../model';
import { parseDate } from '../yearMonth';
import { Statistics } from './statistics';

export interface CategorizationRule {
  readonly pattern: string;
  readonly category: ExpenseCategoryId;
  readonly subscription?: boolean;
}

export interface CategorizationResult {
  readonly category: ExpenseCategoryId;
  readonly subscription: boolean;
  /** Une règle apprise de l'utilisateur prime toujours sur une règle intégrée. */
  readonly source: 'user' | 'builtin';
}

/**
 * Normalisation d'un libellé bancaire.
 *
 * Les relevés arrivent en majuscules, sans accents, truncqués, avec des numéros
 * d'autorisation et des dates collées au nom du marchand : « CB CARREFOUR MARKET 4589
 * 12/03 ». On ramène tout à une forme comparable avant d'appliquer la moindre règle.
 */
export function normalize(label: string): string {
  return label
    .normalize('NFD')
    .replace(/[̀-ͯ]/g, '')
    .toUpperCase()
    .replace(/\b(CB|CARTE|PAIEMENT|ACHAT|VIR|VIREMENT|PRLV|PRELEVEMENT|FACTURE|RETRAIT)\b/g, ' ')
    .replace(/\d{2}[/.-]\d{2}([/.-]\d{2,4})?/g, ' ')
    .replace(/[^A-Z ]/g, ' ')
    .replace(/\s+/g, ' ')
    .trim();
}

/**
 * Règles intégrées.
 *
 * L'ordre compte : « TOTALENERGIES » doit être testé avant « TOTAL », sinon une facture
 * d'électricité serait classée en carburant.
 */
const BUILTIN_RULES: CategorizationRule[] = [
  // Énergie — avant les enseignes dont le nom est un préfixe.
  { pattern: 'TOTALENERGIES', category: 'fixed.electricity' },
  { pattern: 'EDF', category: 'fixed.electricity' },
  { pattern: 'ENGIE', category: 'fixed.gas' },
  { pattern: 'ENI', category: 'fixed.gas' },
  { pattern: 'VEOLIA', category: 'fixed.water' },
  { pattern: 'SUEZ', category: 'fixed.water' },
  { pattern: 'SAUR', category: 'fixed.water' },

  // Courses
  { pattern: 'CARREFOUR', category: 'variable.groceries' },
  { pattern: 'LECLERC', category: 'variable.groceries' },
  { pattern: 'INTERMARCHE', category: 'variable.groceries' },
  { pattern: 'AUCHAN', category: 'variable.groceries' },
  { pattern: 'LIDL', category: 'variable.groceries' },
  { pattern: 'ALDI', category: 'variable.groceries' },
  { pattern: 'CASINO', category: 'variable.groceries' },
  { pattern: 'MONOPRIX', category: 'variable.groceries' },
  { pattern: 'FRANPRIX', category: 'variable.groceries' },
  { pattern: 'SUPER U', category: 'variable.groceries' },
  { pattern: 'HYPER U', category: 'variable.groceries' },
  { pattern: 'CORA', category: 'variable.groceries' },
  { pattern: 'DELHAIZE', category: 'variable.groceries' },
  { pattern: 'COLRUYT', category: 'variable.groceries' },
  { pattern: 'BIOCOOP', category: 'variable.groceries' },
  { pattern: 'PICARD', category: 'variable.groceries' },
  { pattern: 'BOULANGERIE', category: 'variable.groceries' },

  // Restauration
  { pattern: 'MCDONALD', category: 'variable.restaurants' },
  { pattern: 'BURGER KING', category: 'variable.restaurants' },
  { pattern: 'UBER EATS', category: 'variable.restaurants' },
  { pattern: 'DELIVEROO', category: 'variable.restaurants' },
  { pattern: 'JUST EAT', category: 'variable.restaurants' },
  { pattern: 'RESTAURANT', category: 'variable.restaurants' },
  { pattern: 'BRASSERIE', category: 'variable.restaurants' },
  { pattern: 'STARBUCKS', category: 'variable.restaurants' },
  { pattern: 'SUBWAY', category: 'variable.restaurants' },
  { pattern: 'PIZZA', category: 'variable.restaurants' },

  // Carburant et transports
  { pattern: 'TOTAL', category: 'variable.fuel' },
  { pattern: 'ESSO', category: 'variable.fuel' },
  { pattern: 'SHELL', category: 'variable.fuel' },
  { pattern: 'BP ', category: 'variable.fuel' },
  { pattern: 'AVIA', category: 'variable.fuel' },
  { pattern: 'STATION', category: 'variable.fuel' },
  { pattern: 'SNCF', category: 'variable.publicTransport' },
  { pattern: 'RATP', category: 'variable.publicTransport' },
  { pattern: 'NAVIGO', category: 'variable.publicTransport' },
  { pattern: 'UBER', category: 'variable.publicTransport' },
  { pattern: 'BLABLACAR', category: 'variable.publicTransport' },
  { pattern: 'STIB', category: 'variable.publicTransport' },
  { pattern: 'TEC ', category: 'variable.publicTransport' },
  { pattern: 'PEAGE', category: 'variable.fuel' },
  { pattern: 'PARKING', category: 'variable.publicTransport' },

  // Abonnements
  { pattern: 'NETFLIX', category: 'fixed.subscriptions', subscription: true },
  { pattern: 'SPOTIFY', category: 'fixed.subscriptions', subscription: true },
  { pattern: 'DEEZER', category: 'fixed.subscriptions', subscription: true },
  { pattern: 'DISNEY', category: 'fixed.subscriptions', subscription: true },
  { pattern: 'CANAL', category: 'fixed.subscriptions', subscription: true },
  { pattern: 'AMAZON PRIME', category: 'fixed.subscriptions', subscription: true },
  { pattern: 'APPLE COM BILL', category: 'fixed.subscriptions', subscription: true },
  { pattern: 'GOOGLE STORAGE', category: 'fixed.subscriptions', subscription: true },
  { pattern: 'MICROSOFT', category: 'fixed.subscriptions', subscription: true },
  { pattern: 'OPENAI', category: 'fixed.subscriptions', subscription: true },
  { pattern: 'BASIC FIT', category: 'fixed.subscriptions', subscription: true },
  { pattern: 'FITNESS PARK', category: 'fixed.subscriptions', subscription: true },

  // Télécom
  { pattern: 'ORANGE', category: 'fixed.phone' },
  { pattern: 'SFR', category: 'fixed.phone' },
  { pattern: 'BOUYGUES', category: 'fixed.phone' },
  { pattern: 'FREE MOBILE', category: 'fixed.phone' },
  { pattern: 'FREE ', category: 'fixed.internet' },
  { pattern: 'PROXIMUS', category: 'fixed.phone' },
  { pattern: 'VOO', category: 'fixed.internet' },

  // Logement et assurances
  { pattern: 'LOYER', category: 'fixed.rent' },
  { pattern: 'AXA', category: 'fixed.insurance' },
  { pattern: 'MAIF', category: 'fixed.insurance' },
  { pattern: 'MACIF', category: 'fixed.insurance' },
  { pattern: 'MATMUT', category: 'fixed.insurance' },
  { pattern: 'ALLIANZ', category: 'fixed.insurance' },
  { pattern: 'GENERALI', category: 'fixed.insurance' },
  { pattern: 'ASSURANCE', category: 'fixed.insurance' },
  { pattern: 'MUTUELLE', category: 'fixed.insurance' },
  { pattern: 'IMPOT', category: 'fixed.taxes' },
  { pattern: 'DGFIP', category: 'fixed.taxes' },
  { pattern: 'TRESOR PUBLIC', category: 'fixed.taxes' },

  // Santé
  { pattern: 'PHARMACIE', category: 'variable.health' },
  { pattern: 'DOCTEUR', category: 'variable.health' },
  { pattern: 'CABINET MEDICAL', category: 'variable.health' },
  { pattern: 'LABORATOIRE', category: 'variable.health' },
  { pattern: 'DENTISTE', category: 'variable.health' },
  { pattern: 'OPTIC', category: 'variable.health' },

  // Achats divers
  { pattern: 'AMAZON', category: 'variable.household' },
  { pattern: 'FNAC', category: 'variable.leisure' },
  { pattern: 'DECATHLON', category: 'variable.leisure' },
  { pattern: 'CULTURA', category: 'variable.leisure' },
  { pattern: 'CINEMA', category: 'variable.leisure' },
  { pattern: 'UGC', category: 'variable.leisure' },
  { pattern: 'ZARA', category: 'variable.clothing' },
  { pattern: 'H M', category: 'variable.clothing' },
  { pattern: 'UNIQLO', category: 'variable.clothing' },
  { pattern: 'KIABI', category: 'variable.clothing' },
  { pattern: 'IKEA', category: 'variable.household' },
  { pattern: 'LEROY MERLIN', category: 'variable.household' },
  { pattern: 'CASTORAMA', category: 'variable.household' },
  { pattern: 'BRICO', category: 'variable.household' },

  // Enfants et animaux
  { pattern: 'CRECHE', category: 'variable.children' },
  { pattern: 'CENTRE DE LOISIRS', category: 'variable.children' },
  { pattern: 'CANTINE', category: 'variable.children' },
  { pattern: 'VETERINAIRE', category: 'variable.pets' },
  { pattern: 'MAXI ZOO', category: 'variable.pets' },
  { pattern: 'ANIMALERIE', category: 'variable.pets' },

  // Soins personnels
  { pattern: 'COIFFEUR', category: 'variable.personal' },
  { pattern: 'SEPHORA', category: 'variable.personal' },
  { pattern: 'NOCIBE', category: 'variable.personal' },
  { pattern: 'YVES ROCHER', category: 'variable.personal' },
];

/**
 * Catégorisation d'un libellé.
 *
 * Entièrement locale : aucun libellé ne quitte la machine. Un relevé bancaire est une
 * des données les plus intimes qui soient — il révèle la santé, les convictions, les
 * habitudes — et n'a aucune raison d'être envoyé ailleurs pour être trié.
 */
export function categorize(
  label: string,
  userRules: readonly CategorizationRule[] = [],
): CategorizationResult | null {
  const normalized = normalize(label);
  if (normalized.length < 2) return null;

  for (const rule of userRules) {
    if (normalized.includes(normalize(rule.pattern))) {
      return { category: rule.category, subscription: rule.subscription ?? false, source: 'user' };
    }
  }

  for (const rule of BUILTIN_RULES) {
    if (normalized.includes(rule.pattern)) {
      return { category: rule.category, subscription: rule.subscription ?? false, source: 'builtin' };
    }
  }

  return null;
}

export interface DetectedRecurrence {
  readonly label: string;
  readonly averageAmount: Money;
  readonly occurrences: number;
  readonly suggestedFrequency: Frequency;
}

/**
 * Détection des charges récurrentes cachées dans les transactions.
 *
 * Une dépense qui revient à intervalle régulier n'est pas une dépense variable : la
 * traiter comme telle fausse à la fois la projection du mois et le calcul du disponible.
 * On repère les libellés récurrents pour proposer de les convertir en charge déclarée.
 */
export function detectRecurrences(transactions: readonly Transaction[]): DetectedRecurrence[] {
  const groups = new Map<string, Transaction[]>();

  for (const transaction of transactions) {
    if (transaction.kind !== 'expense') continue;
    const key = normalize(transaction.label);
    if (key.length < 3) continue;
    groups.set(key, [...(groups.get(key) ?? []), transaction]);
  }

  const results: DetectedRecurrence[] = [];

  for (const [label, entries] of groups) {
    if (entries.length < 3) continue;

    const dates = entries.map((entry) => parseDate(entry.date).getTime()).sort((a, b) => a - b);
    const intervals: number[] = [];
    for (let index = 1; index < dates.length; index += 1) {
      intervals.push((dates[index]! - dates[index - 1]!) / 86_400_000);
    }
    if (intervals.length === 0) continue;

    const averageInterval = intervals.reduce((total, value) => total + value, 0) / intervals.length;
    const spread = Math.max(...intervals) - Math.min(...intervals);

    // Un écart trop large entre les intervalles trahit une coïncidence, pas un abonnement.
    if (spread > averageInterval * 0.5) continue;

    const frequency = frequencyFor(averageInterval);
    if (!frequency) continue;

    results.push({
      label,
      averageAmount: Statistics.median(
        entries.map((entry) => entry.amount),
        entries[0]!.amount.currency,
      ),
      occurrences: entries.length,
      suggestedFrequency: frequency,
    });
  }

  return results.sort((a, b) => b.occurrences - a.occurrences);
}

function frequencyFor(days: number): Frequency | null {
  if (days >= 6 && days <= 8) return 'weekly';
  if (days >= 13 && days <= 16) return 'biweekly';
  if (days >= 26 && days <= 35) return 'monthly';
  if (days >= 85 && days <= 95) return 'quarterly';
  if (days >= 175 && days <= 190) return 'semiannual';
  if (days >= 355 && days <= 375) return 'annual';
  return null;
}
